design/README.md

```diff
@@ -0,0 +1,298 @@
+    2024ServerDID --> Mar2024ServerDID
+
+    RootServerDID --> ...
+```
+
+Here is one such flow, with a cold server DID and monthly hot server DID:
+
+``` mermaid
+sequenceDiagram
+    participant AccountDID
+    participant ServerRootDID
+    participant JanServerDID
+    participant FebServerDID
+    actor UserRootDID
+
+    Note over AccountDID, UserRootDID: Earlier Server Setup
+    ServerRootDID ->> JanServerDID: UcanDelegate(any)
+    ServerRootDID ->> FebServerDID: UcanDelegate(any)
+
+    Note over AccountDID, UserRootDID: Account Creation
+    UserRootDID ->>+ JanServerDID: newAccount(RootDID)
+        JanServerDID ->>+ AccountDID: spawn_with_parent!(ServerRootDID)
+        AccountDID ->> ServerRootDID: UcanDelegate(Account) [Note: because parenthood]
+        AccountDID -->> AccountDID: dropPrivateKey!
+        ServerRootDID -->> JanServerDID: [recall that this UCAN delegation exists]
+    JanServerDID ->>- UserRootDID: UcanDelegate(Account)
+
+    Note over AccountDID, UserRootDID: Later Access
+    UserRootDID ->>+ AccountDID: Invoke!
+    AccountDID ->>- UserRootDID: Ok, done
+
+    Note over AccountDID, UserRootDID: JanServerDID goes Byzantine
+    ServerRootDID -x JanServerDID: Revoke!
+    ServerRootDID -->> FebServerDID: [recall that this UCAN delegation exists]
+    FebServerDID ->> UserRootDID: UcanDelegate(Account) [Reissue]
+```
+
+Note that in the revocation case, `RootServerDID` revokes all previous delegations affected by the Byzantine DID. This means that they "should" be re-issued, but the blast radius is contained. If more granularity is desired, the predelegation tree of server agents can be made deeper.
+
+# Entities
+
+```mermaid
+erDiagram
+    FullAccount ||--o{ SubAccount: has
+
+    FullAccount ||--o{ Capability: has
+    SubAccount  ||--o{ Capability: has
+
+    FullAccount ||--o{ App : owns
+
+    App ||--|{ HostName : has
+    App ||--|| Storage  : has
+
+    UCAN       ||--|{ Capability : contains
+    Revocation ||--|| UCAN       : revokes
+
+    FullAccount {
+        string   did            PK
+        string   username       UK
+        string   email
+    }
+
+    SubAccount {
+        string   did            PK
+        int      manager_id     FK
+    }
+
+    HostName {
+      string     did            PK
+      string     domain_name
+      string     subdomain
+    }
+
+    App {
+      string     did            PK
+      string     owner_id       FK
+      string     host_name_id   FK
+      string     storage_id     FK
+    }
+
+    Capability {
+        int      id             PK
+        string   cid
+        string   aud            FK
+        string   uri
+        string   ability
+        string   caveat
+        datetime nbf
+        datetime exp
+    }
+
+    Revocation {
+        string   revocation_cid PK
+        string   ucan_cid
+    }
+```
+
+# API Endpoints
+
+While the Fission Server may eventually switch to a RPC design, today we use a JSON REST API:
```

Comment:
Talked a bit to @hugomrdias today. Actually RPC sounds really good.
Looking into it deeper though, I don't think the difference between REST and RPC is that big. Both are just some encoding of actions for http requests.
To my understanding, in REST you tend to send the arguments using path segments, query params and choose the function through HTTP path + HTTP verb, whereas with REST you tend to choose the function through mostly only the HTTP path, and provide the arguments in the body.

The reason we like RPC is because it's easy to "name" the function in a capability. We could have a capability like `"https://runfission.com/api/v0/create-account": { "http/post": [{}] }`.
If we use REST, it's somewhat harder to "name" the function that would e.g. create an arbitrary sub-account due to the path segment in between:
`"https://runfission.com/api/v0/account/???/sub": { "http/post": [{}] }`.
However, it's still possible, we just need to agree on some format, e.g. using asterisks `*`.

Anyway. The answer to RPC or REST may just end up being "both": REST when we have CRUD stuff, RPC when we have action-like stuff.

Reply:
> If we use REST, it's somewhat harder to "name" the function that would e.g. create an arbitrary sub-account due to the path segment in between:
"https://runfission.com/api/v0/account/???/sub": { "http/post": [{}] }.

I may be misunderstanding the question, so let me know if I'm talking past you. I wouldn't model endpoints directly for our use case. The thing you're delegating is a specific URI (for us, usually a DID). _Creating accounts is always allowed_ in the currnet design, which is the same functionality as you see with OAuth flows and whatnot.

For argument sake, what _would_ that look like? 👇

```
"did:key:zSomeAccount": { "child/create": [{}] }
```